scipy/sparse/csgraph/_flow.pyx

```diff
@@ -670,3 +683,943 @@ cdef ITYPE_t[:] _dinic(
+    elif cost.dtype != ITYPE:
+        cost = cost.astype(ITYPE)
+
+    n_verts = csgraph.indptr.shape[0] - 1
+
+    _network_simplex_checks(csgraph.data, demand, cost)
+    tails = _make_tails(csgraph)
+    row = np.empty(cost.shape, dtype=ITYPE)
+    col = np.empty(cost.shape, dtype=ITYPE)
+    flow_data = np.empty(cost.shape, dtype=ITYPE)
+    ns_result = _network_simplex(csgraph.indices, tails, csgraph.data,
+                                 demand, cost, n_verts,
+                                 row, col, flow_data)
+    if not ns_result.is_correct:
+        raise ValueError("no flow satisfies all node demands")
+    flow_matrix = csr_matrix((flow_data[0:ns_result.size],
+                             (row[0:ns_result.size], col[0:ns_result.size])),
+                             shape=(n_verts, n_verts))
+    return MinimumCostFlowResult(flow_matrix, ns_result.flow_cost)
+
+
+def _network_simplex_checks(
+        capacities,  # IN
+        demand,  # IN
+        cost,  # IN
+        ):
+    """
+    Performs checks on input parameter to make sure that they
+    comply with the values of the network simplex algorithms.
+    """
+    if np.any(demand == INT_MAX) or np.any(demand == INT_MIN):
+        raise ValueError("one of the vertices has infinite demand")
+
+    if np.any(cost == INT_MAX) or np.any(cost == INT_MIN):
+        raise ValueError("one of the edges has infinite cost")
+
+    if np.any(capacities < 0):
+        raise ValueError("one of the edges has negative capacity")
+
+    demand_sum = demand.sum()
+
+    if demand_sum != 0:
+        raise ValueError("sum of demands is not zero")
+
+
+cdef void _initialize_spanning_tree(
+        ITYPE_t n_verts,  # IN
+        ITYPE_t n_non_zero_edges,  # IN
+        ITYPE_t faux_inf,  # IN
+        ITYPE_t[:] demand,  # IN
+        ITYPE_t[:] edge_flow,  # IN/OUT
+        ITYPE_t[:] vertex_potentials,  # IN/OUT
+        ITYPE_t[:] parent,  # IN/OUT
+        ITYPE_t[:] parent_edge,  # IN/OUT
+        ITYPE_t[:] subtree_size,  # IN/OUT
+        ITYPE_t[:] next_vertex_dft,  # IN/OUT
+        ITYPE_t[:] prev_vertex_dft,  # IN/OUT
+        ITYPE_t[:] last_descendent_dft  # IN/OUT
+        ) nogil:
+    """
+    This function initialises a spanning tree
+    which can be a feasible solution to the input minimum cost
+    flow problem. The _network_simplex function then performs
+    the algorithm to iteratively update this tree until an 
+    optimal solution is achieved.
+    """
+    cdef ITYPE_t e, v
+
+    # edge_flow initialization
+    for e in range(n_non_zero_edges):
+        edge_flow[e] = 0
+    for v in range(n_verts):
+        edge_flow[v + n_non_zero_edges] = abs(demand[v])
+
+    # vertex_potentials initialization
+    for v in range(1, n_verts + 1):
+        if demand[v - 1] <= 0:
+            vertex_potentials[v] = faux_inf
+        else:
+            vertex_potentials[v] = -faux_inf
+
+    # parent initialization
+    parent[0] = -2
```

Comment:
IIUC, `-2` is like a placeholder values to specify that a node has no parent: is this right?

If it is the case, would it be worth defining a global constant for it?